libs/vkd3d/command.c

```diff
@@ -7720,10 +7720,90 @@ struct d3d12_image_copy_barrier
     bool dst_is_depth_stencil;
 };
 
+static bool d3d12_command_list_check_ds_color_copy_compatibility_24_32bit(const struct vkd3d_format *color_format)
+{
+    switch (color_format->vk_format) {
+        case VK_FORMAT_R32_SFLOAT:
+        case VK_FORMAT_R32_SINT:
+        case VK_FORMAT_R32_UINT:
+            return true;
+        default:
+            return false;
+    }
+}
+
+static bool d3d12_command_list_check_ds_color_copy_compatibility_16bit(const struct vkd3d_format *color_format)
+{
+    switch (color_format->vk_format) {
+        case VK_FORMAT_R16_SFLOAT:
+        case VK_FORMAT_R16_UNORM:
+        case VK_FORMAT_R16_SNORM:
+        case VK_FORMAT_R16_UINT:
+        case VK_FORMAT_R16_SINT:
+            return true;
+        default:
+            return false;
+    }
+}
+
+static bool d3d12_command_list_check_ds_color_copy_compatibility_8bit(const struct vkd3d_format *color_format)
+{
+    switch (color_format->vk_format) {
+        case VK_FORMAT_R8_UINT:
+        case VK_FORMAT_R8_SINT:
+        case VK_FORMAT_R8_UNORM:
+        case VK_FORMAT_R8_SNORM:
+            return true;
+        default:
+            return false;
+    }
+}
+
+/* maintenance8 enables copies between:
+    • 32-bit depth (VK_FORMAT_D32_SFLOAT, VK_FORMAT_D32_SFLOAT_S8_UINT)
+        ◦ VK_FORMAT_R32_SFLOAT, VK_FORMAT_R32_SINT, VK_FORMAT_R32_UINT
+    • 24-bit depth (VK_FORMAT_X8_D24_UNORM_PACK32, VK_FORMAT_D24_UNORM_S8_UINT)
+        ◦ VK_FORMAT_R32_SFLOAT, VK_FORMAT_R32_SINT, VK_FORMAT_R32_UINT
+    • 16-bit depth (VK_FORMAT_D16_UNORM, VK_FORMAT_D16_UNORM_S8_UINT)
+        ◦ VK_FORMAT_R16_SFLOAT, VK_FORMAT_R16_UNORM, VK_FORMAT_R16_SNORM, VK_FORMAT_R16_UINT, VK_FORMAT_R16_SINT
+    • 8-bit stencil (VK_FORMAT_S8_UINT, VK_FORMAT_D16_UNORM_S8_UINT, VK_FORMAT_D24_UNORM_S8_UINT, VK_FORMAT_D32_SFLOAT_S8_UINT)
+        ◦ VK_FORMAT_R8_UINT, VK_FORMAT_R8_SINT, VK_FORMAT_R8_UNORM, VK_FORMAT_R8_SNORM
+ */
+static bool d3d12_command_list_check_ds_color_copy_compatibility(const struct vkd3d_format *dst_format, const struct vkd3d_format *src_format)
+{
+    VkImageAspectFlags ds_bits = VK_IMAGE_ASPECT_DEPTH_BIT | VK_IMAGE_ASPECT_STENCIL_BIT;
+    const struct vkd3d_format *ds_format = dst_format->vk_aspect_mask & ds_bits ? dst_format : src_format;
+    const struct vkd3d_format *color_format = ds_format == dst_format ? src_format : dst_format;
+
+    /* ensure formats detected as expected */
+    if ((ds_format->vk_aspect_mask & ds_bits) == 0 ||
+        (color_format->vk_aspect_mask & VK_IMAGE_ASPECT_COLOR_BIT) == 0)
+        return false;
+
+    /* try the stencil copy case first: always 8bit */
+    if (ds_format->vk_aspect_mask & VK_IMAGE_ASPECT_STENCIL_BIT &&
+        color_format->byte_count == 1)
+        return d3d12_command_list_check_ds_color_copy_compatibility_8bit(color_format);
+
+    switch (ds_format->vk_format)
```

Comment:
Explicitly checking formats is overzealous. The list in maintenance8 is meant to be exhaustive to any possible format D3D12 could use. I think this can be just simplified to "one is color, another is depth-stencil".

Reply:
The z is for zealous.